12527 move staticroutes to bundle

mattamon opened the pull request:
<!--

Before working on a contribution, you must determine on which branch you need to work:
- Bug fix: choose the latest maintenance branch `10.6`
- Feature/Improvement: choose `11.x` 

> All bug fixes merged into the latest maintenance branch are also merged to the current dev branch (`11.x`) on a regular basis.

## Please make sure your PR complies with all of the following points: 
- [ ] Read and accept our [contributing guidelines](/CONTRIBUTING.md) before you submit a PR.
- [ ] Features need to be proper documented in `doc/` 
- [ ] Bugfixes need a short guide how to reproduce them -> target branch is the oldest supported maintenance branch, e.g. `10.6` (see Readme.md for the list of supported versions)
- [ ] Meet all coding standards (see PhpStan actions) 

**Don't submit a PR if it doesn't comply, it'll be closed without a comment!**
-->  
  

## Changes in this pull request  
Resolves #12527 

## Additional info  
Moving static routes functionality into StaticRoutesBundle
Able to install and uninstall via Installer. Permissions are in the Bundle
Moved translations into the bundle

mattamon committed:
Move static routes to bundle

mattamon committed:
Added installer
Added icons
Added migration
Added translation

mattamon committed:
Removing Pimcore\Model\Staticroutes

dvesh3 commented on bundles/AdminBundle/src/Controller/Admin/IndexController.php:
```diff
@@ -22,6 +22,7 @@
 use Pimcore\Bundle\AdminBundle\Controller\AdminController;
 use Pimcore\Bundle\AdminBundle\HttpFoundation\JsonResponse;
 use Pimcore\Bundle\AdminBundle\Security\CsrfProtectionHandler;
+use Pimcore\Bundle\StaticRoutesBundle\Model\Staticroute;
```
refactor the [pimcore settings](https://github.com/pimcore/pimcore/blob/77a80f1a3c61bee064d15fcc1ece929ed0316904/bundles/AdminBundle/src/Controller/Admin/IndexController.php#L249) to bundle as this is just required in `staticroutes.js`

dvesh3 commented on bundles/CoreBundle/src/EventListener/Frontend/ElementListener.php:
```diff
@@ -17,6 +17,7 @@
 
 use Pimcore\Bundle\AdminBundle\Security\User\UserLoader;
 use Pimcore\Bundle\CoreBundle\EventListener\Traits\PimcoreContextAwareTrait;
+use Pimcore\Bundle\StaticRoutesBundle\Model\Staticroute;
```
how does it work when the bundle is disabled?

mattamon replied:
Maybe we can talk about this, because I am not sure how to refactor this method
```php
  protected function applyTargetGroups(Request $request, Document $document)
    {
        if (!$document instanceof Document\Targeting\TargetingDocumentInterface || null !== Staticroute::getCurrentRoute()) {
            return;
        }

        // reset because of preview and editmode (saved in session)
        $document->setUseTargetGroup(null);

        $this->targetingConfigurator->configureTargetGroup($document);

        if ($document->getUseTargetGroup()) {
            $this->logger->info('Setting target group to {targetGroup} for document {document}', [
                'targetGroup' => $document->getUseTargetGroup(),
                'document' => $document->getFullPath(),
            ]);
        }
    }
```

brusch replied:
I think here we need to check whether the targeting bundle is installed or not, maybe just by using a `class_exists()`.

dvesh3 commented on bundles/CoreBundle/src/EventListener/Frontend/HardlinkCanonicalListener.php:
```diff
@@ -17,12 +17,12 @@
 namespace Pimcore\Bundle\CoreBundle\EventListener\Frontend;
 
 use Pimcore\Bundle\CoreBundle\EventListener\Traits\PimcoreContextAwareTrait;
+use Pimcore\Bundle\StaticRoutesBundle\Model\Staticroute;
```
see [ElementListener.php](https://github.com/pimcore/pimcore/pull/13883/files#diff-57746a49d495fb4b2e7c5f5d234b911980876988e503b33427ec5308c8258058)

mattamon replied:
Same here I am not sure how to remove this from the Listener and what would be the implications if I moved this to the bundle.

dvesh3 commented on bundles/StaticRoutesBundle/src/Migrations/Version20221222134837.php:
```diff
@@ -0,0 +1,44 @@
+ *  @license    http://www.pimcore.org/license     GPLv3 and PCL
+ */
+
+namespace Pimcore\Bundle\StaticRoutesBundle\Migrations;
```
move to core bundle

dvesh3 commented on bundles/StaticRoutesBundle/src/Migrations/Version20221222134837.php:
```diff
@@ -0,0 +1,44 @@
+use Pimcore\Model\Tool\SettingsStore;
+
+/**
+ * Staticroutes will be enabled by default
```
```suggestion
 * Staticroutes will be installed by default
```

dvesh3 replied:
```suggestion
 * Staticroutes will be installed by default
```

dvesh3 commented on lib/Targeting/EventListener/DocumentTargetGroupListener.php:
```diff
@@ -17,12 +17,12 @@
 
 namespace Pimcore\Targeting\EventListener;
 
+use Pimcore\Bundle\StaticRoutesBundle\Model\Staticroute;
```
see [ElementListener.php](https://github.com/pimcore/pimcore/pull/13883/files#diff-57746a49d495fb4b2e7c5f5d234b911980876988e503b33427ec5308c8258058)

mattamon replied:
Same as the other Listeners

dvesh3 commented on bundles/StaticRoutesBundle/src/StaticRoutesBundle.php:
```diff
@@ -0,0 +1,54 @@
+use Pimcore\Extension\Bundle\AbstractPimcoreBundle;
+use Pimcore\Extension\Bundle\Traits\PackageVersionTrait;
+
+class StaticRoutesBundle extends AbstractPimcoreBundle
```
rename to `PimcoreStaticRoutesBundle`

mattamon committed:
Removing translations
Rename bundle from StaticRoutesBundle to PimcoreStaticRoutesBundle
Adding Eventlistener for pimcore settings
Moving migration to core

mattamon committed:
Add class exists checks for static routes

mattamon committed:
Merge branch '11.x' into 12527-move-staticroutes-to-bundle

mattamon committed:
Adding strict to installer

mattamon committed:
Merge branch '11.x' into 12527-move-staticroutes-to-bundle

# Conflicts:
#	.github/ci/files/config/bundles.php
#	bundles/StaticRoutesBundle/src/Routing/Staticroute/Router.php
#	composer.json

mattamon committed:
Added independant configuration
Updated namespaces and routes
Added upgrade notes

mattamon committed:
Refactor JS namespaces and extracting translations to admin.en.extended.json

mattamon committed:
Merge branch '11.x' into 12527-move-staticroutes-to-bundle

mattamon committed:
Merge branch '11.x' into 12527-move-staticroutes-to-bundle

# Conflicts:
#	bundles/StaticRoutesBundle/public/js/settings.js

mattamon committed:
Javascript: tag private classes

dvesh3 commented on bundles/CoreBundle/src/EventListener/Frontend/ElementListener.php:
```diff
@@ -130,7 +130,11 @@ protected function handleVersion(Request $request, Document $document): Document
+            return;
+        }
+
+        if(class_exists('Pimcore\Bundle\StaticRoutesBundle\Model\Staticroute') && null !== Staticroute::getCurrentRoute()) {
```
```suggestion
        if(class_exists(Staticroute::class) && null !== Staticroute::getCurrentRoute()) {
```

dvesh3 commented on bundles/CoreBundle/src/EventListener/Frontend/HardlinkCanonicalListener.php:
```diff
@@ -70,7 +70,11 @@ public function onKernelResponse(ResponseEvent $event)
+            $this->handleHardlink($request, $event->getResponse(), $document);
+        }
+
+        if ($document instanceof WrapperInterface) {
```
```suggestion
        if (class_exists(Staticroute::class) && null !== Staticroute::getCurrentRoute()) {
            return;
        }

        if ($document instanceof WrapperInterface) {
```

dvesh3 replied:
this can also be moved above document resolver.

dvesh3 commented on bundles/CoreBundle/src/Migrations/Version20221222134837.php:
```diff
@@ -0,0 +1,43 @@
+
+        $this->warnIf(
+            null !== SettingsStore::get('BUNDLE_INSTALLED__Pimcore\\Bundle\\StaticRoutesBundle\\StaticRoutesBundle', 'pimcore'),
+            'Please make sure to enable the Pimcore\\Bundle\\StaticRoutesBundle\\StaticRoutesBundle manually in config/bundles.php'
```
```suggestion
            null !== SettingsStore::get('BUNDLE_INSTALLED__Pimcore\\Bundle\\StaticRoutesBundle\\PimcoreStaticRoutesBundle', 'pimcore'),
            'Please make sure to enable the Pimcore\\Bundle\\StaticRoutesBundle\\PimcoreStaticRoutesBundle manually in config/bundles.php'
```

dvesh3 commented on lib/Targeting/EventListener/DocumentTargetGroupListener.php:
```diff
@@ -74,7 +74,11 @@ public function onVisitorInfoResolve(TargetingEvent $event)
+            return;
+        }
+
+        if(class_exists('Pimcore\Bundle\StaticRoutesBundle\Model\Staticroute') && null !== Staticroute::getCurrentRoute()) {
```
```suggestion
        if(class_exists(Staticroute::class) && null !== Staticroute::getCurrentRoute()) {
```

mattamon committed:
Updating preprareDataStructureForYaml to new structure

mattamon committed:
Update upgrade notes for new config

mattamon committed:
Replacing hardcoded class with class constant
Updating migration to the correct bundle namespace

dvesh3 commented on bundles/CoreBundle/src/Migrations/Version20221222134837.php:
```diff
@@ -0,0 +1,43 @@
+        return 'Install static routes bundle by default';
+    }
+
+    public function up(Schema $schema): void
```
please implement a down method which does nothing...the downgrade migration throws error without `down` method implementation:
![image](https://user-images.githubusercontent.com/5137917/211537146-276d0ed6-dfa2-4813-ad52-7a6f7c2d7ac6.png)

mattamon committed:
Merge branch '11.x' into 12527-move-staticroutes-to-bundle

# Conflicts:
#	.github/ci/files/config/bundles.php
#	bundles/AdminBundle/src/Controller/Admin/SettingsController.php
#	composer.json
#	doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md

mattamon committed:
Add down method to migration

mattamon committed:
Merge branch '11.x' into 12527-move-staticroutes-to-bundle

# Conflicts:
#	bundles/CoreBundle/translations/en.extended.json

mattamon committed:
Change the permission to routes and add a category
Add the translation for routes

mattamon committed:
Move extended.json translations to admin.en.yaml

mattamon committed:
Merge branch '11.x' into 12527-move-staticroutes-to-bundle

mattamon committed:
Merge branch '11.x' into 12527-move-staticroutes-to-bundle

# Conflicts:
#	bundles/StaticRoutesBundle/src/Routing/Staticroute/Router.php
#	composer.json
#	lib/Routing/Element/Router.php

mattamon committed:
Add native return types
Remove unnecessary if conditions

mattamon committed:
Merge branch '11.x' into 12527-move-staticroutes-to-bundle

# Conflicts:
#	.github/ci/files/config/bundles.php

dvesh3 reviewed (dismissed):
Move Kernel Config https://github.com/pimcore/pimcore/blob/1ce2021a00e338c4548fb38c7a3e2ab716ec55d6/lib/Kernel.php#L171

mattamon committed:
Merge branch '11.x' into 12527-move-staticroutes-to-bundle
